[MERGE] tools, base, *: improve image tools

seb-odoo opened the pull request:
Merge commit for PR: #31811 

The main goal is to keep the image ratio on resize (instead of adding a white border).

The opportunity is taken to also clean up the image tools.

See individual commit messages for more information.

Functional:

- Remove the white border from all images (never resize to square, keep aspect ratio)
- Prevent some images from being 3px shorter than expected
- Resize the website favicons appropriately, always encode them as ICO, add a default favicon to new websites
- Secure images going through the tools by returning an error when non-image content is given
- Incrase the max image filesize from 42e6 to 45e6 to fit some of the biggest images we might get from Unsplash
- Always return the correct mimetype/Content-Type for images
- The avatar placeholder was white displayed on top of white, now it's grey so we can actually see it

Technical:

- make image tools more consistent by merging the tools and removing duplicate code
- add tests for the image tools
- remove uncessary parameters
- remove the "ghost border fix"


task-1958000

pedrobaeza commented:
Today in an issue has risen the no support for Webp image format. Although we depend in the Imagemagick library on background and I'm not sure if it's supported, do you think it's feasible?

pedrobaeza commented:
And another thing :angel: : is SVG format still supported with these changes?

seb-odoo commented:
@pedrobaeza I made an analysis of webp at the start of the task, and the short answer is no because we would need to support both webp and not webp version of the image, this is too much complexity for the gain.

About WebP
==========

Compression

For a similar file size than jpeg, it creates a better looking image.
The loss of quality happens by lost of details and added blurriness, which are more pleasant to the eyes than the artifacts of jpeg.
- Image in quality 80 is almost like the original, for a bit lower size than the jpeg in 80.
- Image in quality 50 looks honestly good, for a low filesize. jpeg doesn't even come close in quality for that size.
- Image in quality 0 looks very decent for how small the file is. This is ideal for "ugly" placeholders. jpeg in quality 0 has the same filesize but the quality is crap.
Lossless is not good at all. If the file has been encoded in jpeg, the jpeg is going to be smaller than the webp lossless.


Compatibility

WebP is a good image format, unfortunately it is not supported on all browsers, especially not IE and Safari.
However, the browsers that do support it include "image/webp" in the http header Accept when requesting an image, so it would be possible to create a fallback.
This adds complexity in the code, for what gain?
- If we are going to use quality 80, I would say it's not worth the effort. However, if we want to work with quality 50 or less, the impact on quality is significant.


Performance (this applies to jpeg too):

Resizing or reencoding the image in real time is slow and CPU heavy. For a really big original image, it can take around 2 seconds.
We need to precompute and store various intermediary sizes and serve those.
Or make a smart image cache system.
Or we need to somehow rely on CDN but that would make Odoo barely usable without it.


Security

The WebP Plugin for Pillow as well as libwebp have been reveiwed by ODO and are considered safe enough for us to use.


Dependency
libwebp is already a requirement of Pillow so it should be present by default on all systems running Odoo.

seb-odoo commented:
@pedrobaeza SVG is supported and will keep being supported, but there is an ORM security restriction that only the admin can upload them due to the possibility to inject JS on them.

pedrobaeza commented:
Thanks for the info about WebP and SVG, Sébastien!

My interest about SVG is because I was the one that added that support in v12 :smiley:

KangOl commented on odoo/tools/image.py:
```diff
@@ -26,132 +27,113 @@
+            if new_h > h:
+                new_w, new_h = (new_w * h) // new_h, h
+
+            if type == 'top':
```
always false. `type` is the builtin function https://docs.python.org/3.5/library/functions.html#type

seb-odoo replied:
Good catch, the code was coming from another function where type was shadowed by a local variable.

rdeodoo commented on addons/website_profile/controllers/main.py:
```diff
@@ -30,18 +31,10 @@ def _check_avatar_access(self, user_id, **post):
-            if suffix in ('small', 'medium', 'large', 'big'):
-                content = getattr(tools, 'image_resize_image_%s' % suffix)(content)
-        return content
+            return base64.b64encode(f.read())
```
Particular reason why is the 'resize guess based on name' got moved outside the `_get_default_avatar`?

seb-odoo replied:
It was copy pasted there from another place, so I moved it to a central place to remove duplication.

rdeodoo commented on odoo/tools/image.py:
```diff
@@ -26,132 +28,116 @@
+        - The original image ratio is preserved, unless `crop` is also given.
+        - If width or height is falsy, it will be computed from the other value
+            and from the ratio of the original image.
+        - If size is falsy or both with and height are falsy, no resize is done.
```
both wi**d**th and height

seb-odoo replied:
Good catch!

rdeodoo commented on addons/web_unsplash/controllers/main.py:
```diff
@@ -89,22 +89,18 @@ def slugify(s):
-                    datas = tools.image_save_for_web(image)
-                    # append image extension in name
-                    query += '.' + str.lower(image.format)
+            image_base64 = tools.image_process(image_base64)
```
That unsplash code (double mimetype check) was subject to debate and done for a particular reason but I can't remember why, I should have commented the code, we will see what happen :yum:.

seb-odoo replied:
The new method is making sure the received file is a correct image and the mimetype is guessed accordingly too, so hopefully the change won't break anything.

seb-odoo commented:
For `force_contenttype` it was moved outside of the `if` because it is possible for `image_process` to change the format of the image (bmp -> png, and unsupported format -> jpeg), so the content-type that was pre-computed will be wrong in those cases, this is why I reset it after the processing.

seb-odoo commented:
For the "blurry" images it is because I removed the sharpness, that was responsible of making the /shop images ugly. The image that you say were a bit better before is because the original images of them are indeed a bit blurry, and this is not compensated anymore.

seb-odoo commented:
For the resize encapsulated methods I'm not sure. The previous solution with hard-coded parameters were a bit hard to read and to maintain. If we want to keep such helpers encapsulating the method at least they should just take kwargs, but it's still more code that I'm not sure is useful.

seb-odoo commented:
See also enterprise PR https://github.com/odoo/enterprise/pull/3934

seb-odoo commented:
@KangOl I'm finally done with the changes on this PR, will you review it please?

Note : still waiting on the security fix from AKU for the placeholder route.

seb-odoo commented:
Rebased to all the recent changes in master. Still waiting review. @KangOl @JKE-be

KangOl commented on addons/website_slides/controllers/main.py:
```diff
@@ -506,7 +507,7 @@ def slide_get_pdf_content(self, slide):
     @http.route('/slides/slide/<int:slide_id>/get_image', type='http', auth="public", website=True, sitemap=False)
-    def slide_get_image(self, slide_id, field='image_medium', width=0, height=0, crop=False, avoid_if_small=False, upper_limit=False):
+    def slide_get_image(self, slide_id, field='image_medium', width=0, height=0, crop=False):
         # Protect infographics by limiting access to 256px (large) images
```
exact same code as in `website_profile` module (only the model change), can't code be extract to a helper method?

seb-odoo replied:
I agree, not only those 2 but pretty much all the places where `binary_content` is called are doing very similar things. Probably better for another task I think.

JKE-be commented on addons/web/controllers/main.py:
```diff
@@ -1059,6 +1059,9 @@ def content_image(self, xmlid=None, model='ir.attachment', id=None, field='datas
             image_base64, width=width, height=height, crop=crop)
 
+        if not image_base64:
+            return request.not_found()
```
remove this commit if still useless like discuss previously

seb-odoo replied:
Ok to remove (image_base64 shouldn't be empty with the placeholder, and the code will already raise before that point if there is another problem).

JKE-be commented on odoo/tools/image.py:
```diff
@@ -32,89 +33,122 @@
     # get the format of the original image (must be done before resize)
-    opt = {'format': image.format}
+    format = (format or image.format).upper()
+    if format == 'BMP':
```
rename format to avoid erase: <built-in function format>

JKE-be replied:
output_format ? output

seb-odoo replied:
Ok for `output_format`

JKE-be commented on odoo/tools/image.py:
```diff
@@ -32,89 +33,122 @@
-        image = image.convert("RGB")
-
-    return image_to_base64(image, filetype)
+def image_process(base64_source, size=(0, 0), verify_resolution=False, quality=80, crop=None, colorize=False, format=None):
```
@JKE-be 

hard testing to do

seb-odoo replied:
@JKE-be

hard testing to do

(so you get an email)

seb-odoo commented on addons/website/models/website.py:
```diff
@@ -138,6 +148,8 @@ def create(self, vals):
 
     @api.multi
     def write(self, values):
+        tools.update_field_value_to_ico(values)
```
I just moved it to the website model just below the create/write so it's easy to see what it does without having to navigate multiple files.

It wasn't really necessary in tools, I don't think it will be used by anything else than website.

seb-odoo commented:
Pending merge of https://github.com/odoo/odoo/pull/32460 to get rid of the last commit in the current PR.

seb-odoo commented:
I added the Image tools class with smaller and chainable methods 2 weeks ago, the ORM cache issue was fixed last week, I just rebased on latest master again, I don't plan to make any more change, and runbot this green.

I think it is time we finally merge this @KangOl @JKE-be
cc @odony for security check

seb-odoo commented on odoo/http.py:
```diff
@@ -1617,5 +1617,27 @@ def content_disposition(filename):
+    :rtype: list of tuples (name, value)
+    """
+    dictheaders = dict(headers)
+    dictheaders[name] = value
```
I was reading https://github.com/odoo/odoo/pull/26635/#issuecomment-425043088 and I'm thinking this will be an issue for SVG specifically.

We probably need this function to do nothing if the given header is content-type and the given value is SVG?
Or is this function too low level, and there should be another layer to check the security when called from unsafe places?

@odony

seb-odoo replied:
So creating a new function such as `set_safe_image_content_type` that sets it only if the target type is 'PNG', 'JPEG', 'GIF', 'ICO'?

seb-odoo replied:
I have my suggested change ready if needed. Waiting on approval/pending reviews before force pushing.

seb-odoo commented:
@odony rebase done (community)

odony commented:
@robodoo r+ for real

robodoo commented:
Linked pull request(s) odoo/enterprise#3934 not ready. Linked PRs are not staged until all of them are ready.

robodoo commented:
Because this PR has multiple commits, I need to know how to merge it:

* `merge` to merge directly, using the PR as merge commit message
* `rebase-merge` to rebase and merge, using the PR as merge commit message
* `rebase-ff` to rebase and fast-forward

qsm-odoo commented:
@robodoo rebase-merge I guess :)

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message

robodoo commented:
Merged, thanks!

mart-e commented:
@seb-odoo it looks like the payment module can not be installed with Pillow 6.0 since this commit.
The image `payment/static/img/codensa_easy_credit.png` has transparency and crashes in the `image_process` call of payment.acquirer.
Can you have a look?

seb-odoo commented:
@mart-e the pillow from our requirements.txt is 4.0.0, I don't think that we support the version 6?

mart-e commented:
@seb-odoo we don't support a lower version than 4.0 but we do support all the ones above. Easy to reproduce : 

```
$ sudo pip install pillow --upgrade
$ git checkout df189fd6ab749de47ceba35c547db4f903965016~1  # commit just before
$ <odoo-cmd> -i payment
...  # works fine
$ git checkout df189fd6ab749de47ceba35c547db4f903965016
$ <odoo-cmd> -i payment
...
  File "/home/mat/odoo/odoo/addons/payment/models/payment_acquirer.py", line 535, in write
    vals['image_payment_form'] = image_process(image, size=(45,30))
  File "/home/mat/odoo/odoo/odoo/tools/image.py", line 246, in image_process
    return image.image_base64(quality=quality, output_format=output_format)
  File "/home/mat/odoo/odoo/odoo/tools/image.py", line 126, in image_base64
    return image_to_base64(output_image, **opt)
  File "/home/mat/odoo/odoo/odoo/tools/image.py", line 281, in image_to_base64
    image.save(stream, format=format, **params)
  File "/usr/lib/python3.7/site-packages/PIL/Image.py", line 2007, in save
    save_handler(self, fp, filename)
  File "/usr/lib/python3.7/site-packages/PIL/PngImagePlugin.py", line 859, in _save
    red, green, blue = transparency
$ git checkout HEAD~1
````

seb-odoo commented:
Task 1998639 created

seb-odoo commented:
> Task 1998639 created for the pillow 6 issue.

Should be solved in https://github.com/odoo/odoo/pull/33628
